pylint not finding the tox.ini config

### Bug description

According to documentation: https://pylint.readthedocs.io/en/v3.2.3/user_guide/usage/run.html
`pylint` should be able to find `tox.ini` and use configuration from it.

In my case however it is not doing it. It works if I explicitely ask to use tox.ini with `--rcfile`.


### Configuration

```ini
[tox]
env_list =
    py313
    py312
    py311
    py310
    py39
    py38
    type
    linter
    flake8

[testenv]
deps = -r {toxinidir}/requirements_test.txt
commands = pytest

[testenv:flake8]
deps = -r {toxinidir}/requirements_test.txt
commands = flake8

[flake8]
exclude = .tox,tests
max-line-length = 120


[testenv:type]
deps = -r {toxinidir}/requirements_test.txt
commands = mypy --ignore-missing-imports .

[testenv:pylint]
deps = pylint
commands = pylint *.py

[pylint]
disable = C0114,C0115,C0116
```


### Command used

```shell
pylint fraction.py -v
```


### Pylint output

```python
No config file found, using default configuration
************* Module fraction
fraction.py:1:0: C0114: Missing module docstring (missing-module-docstring)
fraction.py:4:0: C0115: Missing class docstring (missing-class-docstring)
fraction.py:63:4: C0116: Missing function or method docstring (missing-function-docstring)
fraction.py:66:4: C0116: Missing function or method docstring (missing-function-docstring)

---------------------------------------------------------------------------------------------------
Your code has been rated at 9.02/10 (previous run: 9.02/10, +0.00)
Checked 1 files, skipped 1 files
```


### Expected behavior

Same as using with `--rcfile`:
```
$ pylint fraction.py -v --rcfile=tox.ini
Using config file tox.ini

----------------------------------------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 9.02/10, +0.98)
Checked 1 files, skipped 1 files
```

### Pylint version

```shell
pylint 3.2.3
astroid 3.2.2
Python 3.12.3 (main, May 22 2024, 16:56:59) [GCC 13.3.0]
```


### OS / Environment

Gentoo

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9727.bugfix
+++ b/doc/whatsnew/fragments/9727.bugfix
@@ -1,3 +1,5 @@
 Fix a bug where a ``tox.ini`` file with pylint configuration was ignored and it exists in the current directory.
 
+``.cfg`` and ``.ini`` files containing a ``Pylint`` configuration may now use a section named ``[pylint]``. This enhancement impacts the scenario where these file types are used as defaults when they are present and have not been explicitly referred to, using the ``--rcfile`` option.
+
 Closes #9727

--- a/doc/whatsnew/fragments/9727.bugfix
+++ b/doc/whatsnew/fragments/9727.bugfix
@@ -1,0 +1,3 @@
+Fix a bug where a ``tox.ini`` file with pylint configuration was ignored and it exists in the current directory.
+
+Closes #9727

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -61,7 +61,10 @@ def _cfg_has_config(path: Path | str) -> bool:
         parser.read(path, encoding="utf-8")
     except configparser.Error:
         return False
-    return any(section.startswith("pylint.") for section in parser.sections())
+    return any(
+        section == "pylint" or section.startswith("pylint.")
+        for section in parser.sections()
+    )
 
 
 def _yield_default_files() -> Iterator[Path]:

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -22,7 +22,8 @@ RC_NAMES = (
     Path(".pylintrc.toml"),
 )
 PYPROJECT_NAME = Path("pyproject.toml")
-CONFIG_NAMES = (*RC_NAMES, PYPROJECT_NAME, Path("setup.cfg"))
+TOX_NAME = Path("tox.ini")
+CONFIG_NAMES = (*RC_NAMES, PYPROJECT_NAME, Path("setup.cfg"), TOX_NAME)
 
 
 def _find_pyproject() -> Path:
@@ -71,7 +72,7 @@ def _yield_default_files() -> Iterator[Path]:
             if config_name.is_file():
                 if config_name.suffix == ".toml" and not _toml_has_config(config_name):
                     continue
-                if config_name.suffix == ".cfg" and not _cfg_has_config(config_name):
+                if config_name.suffix in (".cfg", ".ini") and not _cfg_has_config(config_name):
                     continue
 
                 yield config_name.resolve()

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -72,7 +72,9 @@ def _yield_default_files() -> Iterator[Path]:
             if config_name.is_file():
                 if config_name.suffix == ".toml" and not _toml_has_config(config_name):
                     continue
-                if config_name.suffix in (".cfg", ".ini") and not _cfg_has_config(config_name):
+                if config_name.suffix in (".cfg", ".ini") and not _cfg_has_config(
+                    config_name
+                ):
                     continue
 
                 yield config_name.resolve()

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -74,9 +74,10 @@ def _yield_default_files() -> Iterator[Path]:
             if config_name.is_file():
                 if config_name.suffix == ".toml" and not _toml_has_config(config_name):
                     continue
-                if config_name.suffix in {".cfg", ".ini"} and not _cfg_or_ini_has_config(
-                    config_name
-                ):
+                if config_name.suffix in {
+                    ".cfg",
+                    ".ini",
+                } and not _cfg_or_ini_has_config(config_name):
                     continue
 
                 yield config_name.resolve()

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -22,8 +22,7 @@ RC_NAMES = (
     Path(".pylintrc.toml"),
 )
 PYPROJECT_NAME = Path("pyproject.toml")
-TOX_NAME = Path("tox.ini")
-CONFIG_NAMES = (*RC_NAMES, PYPROJECT_NAME, Path("setup.cfg"), TOX_NAME)
+CONFIG_NAMES = (*RC_NAMES, PYPROJECT_NAME, Path("setup.cfg"), Path("tox.ini"))
 
 
 def _find_pyproject() -> Path:

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -55,7 +55,7 @@ def _toml_has_config(path: Path | str) -> bool:
     return "pylint" in content.get("tool", [])
 
 
-def _cfg_has_config(path: Path | str) -> bool:
+def _cfg_or_ini_has_config(path: Path | str) -> bool:
     parser = configparser.ConfigParser()
     try:
         parser.read(path, encoding="utf-8")
@@ -74,7 +74,7 @@ def _yield_default_files() -> Iterator[Path]:
             if config_name.is_file():
                 if config_name.suffix == ".toml" and not _toml_has_config(config_name):
                     continue
-                if config_name.suffix in {".cfg", ".ini"} and not _cfg_has_config(
+                if config_name.suffix in {".cfg", ".ini"} and not _cfg_or_ini_has_config(
                     config_name
                 ):
                     continue

--- a/pylint/config/find_default_config_files.py
+++ b/pylint/config/find_default_config_files.py
@@ -72,7 +72,7 @@ def _yield_default_files() -> Iterator[Path]:
             if config_name.is_file():
                 if config_name.suffix == ".toml" and not _toml_has_config(config_name):
                     continue
-                if config_name.suffix in (".cfg", ".ini") and not _cfg_has_config(
+                if config_name.suffix in {".cfg", ".ini"} and not _cfg_has_config(
                     config_name
                 ):
                     continue

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -310,7 +310,7 @@ disable = logging-not-lazy,logging-format-interpolation
 def test_has_config(content: str, expected: bool, tmp_path: Path) -> None:
     """Test that a .cfg file or .ini file has a pylint config."""
     for file_name in ("fake.cfg", "tox.ini"):
-        fake_conf = tmp_path / file_name 
+        fake_conf = tmp_path / file_name
         with open(fake_conf, "w", encoding="utf8") as f:
             f.write(content)
         assert _cfg_has_config(fake_conf) == expected

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -307,12 +307,17 @@ disable = logging-not-lazy,logging-format-interpolation
         ],
     ],
 )
-def test_cfg_has_config(content: str, expected: bool, tmp_path: Path) -> None:
-    """Test that a cfg file has a pylint config."""
+def test_has_config(content: str, expected: bool, tmp_path: Path) -> None:
+    """Test that a .cfg file or .ini file has a pylint config."""
     fake_cfg = tmp_path / "fake.cfg"
     with open(fake_cfg, "w", encoding="utf8") as f:
         f.write(content)
     assert _cfg_has_config(fake_cfg) == expected
+
+    fake_ini = tmp_path / "tox.ini"
+    with open(fake_ini, "w", encoding="utf8") as f:
+        f.write(content)
+    assert _cfg_has_config(fake_ini) == expected
 
 
 def test_non_existent_home() -> None:

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -309,10 +309,11 @@ disable = logging-not-lazy,logging-format-interpolation
 )
 def test_has_config(content: str, expected: bool, tmp_path: Path) -> None:
     """Test that a .cfg file or .ini file has a pylint config."""
-    fake_cfg = tmp_path / "fake.cfg"
-    with open(fake_cfg, "w", encoding="utf8") as f:
-        f.write(content)
-    assert _cfg_has_config(fake_cfg) == expected
+    for file_name in ("fake.cfg", "tox.ini"):
+        fake_conf = tmp_path / file_name 
+        with open(fake_conf, "w", encoding="utf8") as f:
+            f.write(content)
+        assert _cfg_has_config(fake_conf) == expected
 
     fake_ini = tmp_path / "tox.ini"
     with open(fake_ini, "w", encoding="utf8") as f:

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -315,11 +315,6 @@ def test_has_config(content: str, expected: bool, tmp_path: Path) -> None:
             f.write(content)
         assert _cfg_has_config(fake_conf) == expected
 
-    fake_ini = tmp_path / "tox.ini"
-    with open(fake_ini, "w", encoding="utf8") as f:
-        f.write(content)
-    assert _cfg_has_config(fake_ini) == expected
-
 
 def test_non_existent_home() -> None:
     """Test that we handle a non-existent home directory.

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -18,7 +18,7 @@ import pytest
 from pytest import CaptureFixture
 
 from pylint import config, testutils
-from pylint.config.find_default_config_files import _cfg_has_config, _toml_has_config
+from pylint.config.find_default_config_files import _cfg_or_ini_has_config, _toml_has_config
 from pylint.lint.run import Run
 
 
@@ -313,7 +313,7 @@ def test_has_config(content: str, expected: bool, tmp_path: Path) -> None:
         fake_conf = tmp_path / file_name
         with open(fake_conf, "w", encoding="utf8") as f:
             f.write(content)
-        assert _cfg_has_config(fake_conf) == expected
+        assert _cfg_or_ini_has_config(fake_conf) == expected
 
 
 def test_non_existent_home() -> None:

--- a/tests/config/test_find_default_config_files.py
+++ b/tests/config/test_find_default_config_files.py
@@ -18,7 +18,10 @@ import pytest
 from pytest import CaptureFixture
 
 from pylint import config, testutils
-from pylint.config.find_default_config_files import _cfg_or_ini_has_config, _toml_has_config
+from pylint.config.find_default_config_files import (
+    _cfg_or_ini_has_config,
+    _toml_has_config,
+)
 from pylint.lint.run import Run
 
 

